ansible_runner/streaming.py

```diff
@@ -0,0 +1,227 @@
+                settings = {}
+        self.config = MockConfig(settings)
+
+        artifact_dir = kwargs.get('artifact_dir')
+        self.artifact_dir = os.path.abspath(
+            artifact_dir or os.path.join(self.private_data_dir, 'artifacts'))
+
+        self.status_handler = status_handler
+        self.event_handler = event_handler
+        self.artifacts_handler = artifacts_handler
+
+        self.cancel_callback = cancel_callback  # FIXME: unused
+        self.finished_callback = finished_callback
+
+        self.status = "unstarted"
+        self.rc = None
+
+    def status_callback(self, status_data):
+        self.status = status_data['status']
+        if self.status == 'starting':
+            self.config.command = status_data.get('command')
+            self.config.env = status_data.get('env')
+            self.config.cwd = status_data.get('cwd')
+
+        for plugin in ansible_runner.plugins:
+            ansible_runner.plugins[plugin].status_handler(self.config, status_data)
+        if self.status_handler is not None:
+            self.status_handler(status_data, runner_config=self.config)
+
+    def event_callback(self, event_data):
+        full_filename = os.path.join(self.artifact_dir,
+                                     'job_events',
+                                     '{}-{}.json'.format(event_data['counter'],
+                                                         event_data['uuid']))
+        if 'stdout' in event_data:
+            print(event_data['stdout'])
+
+        if self.event_handler is not None:
+            should_write = self.event_handler(event_data)
+        else:
+            should_write = True
+        for plugin in ansible_runner.plugins:
+            ansible_runner.plugins[plugin].event_handler(self.config, event_data)
+        if should_write:
+            with codecs.open(full_filename, 'w', encoding='utf-8') as write_file:
+                os.chmod(full_filename, stat.S_IRUSR | stat.S_IWUSR)
+                json.dump(event_data, write_file)
+
+    def artifacts_callback(self, artifacts_data):
+        buf = io.BytesIO(artifacts_data)
+        with zipfile.ZipFile(buf, 'r') as archive:
+            archive.extractall(path=self.config.artifact_dir)
+
+        if self.artifacts_handler is not None:
+            self.artifacts_handler(self.config.artifact_dir)
+
+    def run(self):
+        job_events_path = os.path.join(self.artifact_dir, 'job_events')
+        if not os.path.exists(job_events_path):
+            os.makedirs(job_events_path, 0o700, exist_ok=True)
+
+        while True:
+            line = self._input.readline()
+            data = json.loads(line)
+
+            if 'status' in data:
+                self.status_callback(data)
+            elif 'zipfile' in data:
+                self.artifacts_callback(self._input.read(data['zipfile']))
+            elif 'eof' in data:
+                break
+            else:
+                self.event_callback(data)
```

Comment:
I would like to spend a little time making this `if-elif-elif` super explicit instead of this duck typing thing.

I would advocate for a single key that denotes the type of runner callback. That will be important to clearly distinguish these transport-layer _callbacks_ from _job event_ recording by the plugin or python interface. In terms of the event lifecycle we will have:

 - An event from the actual Ansible stdout callback, which are "partial" events
 - The combined event, dispatched from the worker to the processor to be used in the callback
   - tagged as a unprocessed runner callback, with the callback type
 - Processed event (processed by the callback) sent into the receiving messaging system
   - sent into a dedicated queue for this type of event, contains relational tags like job / workflow / host
 - Event in database

A purist would go even further to say they should have a version associated with them. I wouldn't follow the textbook advice for event-driven architectures there, but I could see some use for returning the _runner_ version, even if only for a single callback, like the start. At some point, we're going to have problems with non-compliant nodes popping up in the execution plane.